bot/exts/summer_aoc.py

```diff
@@ -0,0 +1,323 @@
+            "day_interval": self.day_interval,
+            "post_time": self.post_time,
+        })
+
+    async def start_event(self) -> None:
+        """Start event by recording state and creating async tasks to post puzzles."""
+        log.info(f"Starting Summer AoC event with {self.year=} {self.current_day=} {self.day_interval=}")
+        sleep_for = time_until(hour=self.post_time)
+        self.wait_task = asyncio.create_task(asyncio.sleep(sleep_for.total_seconds()))
+        log.debug(f"Waiting until {self.post_time}:00 UTC to start Summer AoC loop")
+        await self.wait_task
+
+        self.loop_task = tasks.Loop(
+            self.post_puzzle,
+            seconds=0,
+            minutes=0,
+            hours=(self.day_interval * 24),
+            time=MISSING,
+            count=None,
+            reconnect=True,
+        )
+        self.loop_task.start()
+
+    async def stop_event(self) -> bool:
+        """Cancel any active Summer AoC tasks. Returns whether the event was running."""
+        was_waiting = self.wait_task and not self.wait_task.done()
+        was_looping = self.loop_task and self.loop_task.is_running()
+        if was_waiting and was_looping:
+            log.error("Both wait and loop tasks were active. Both should now be cancelled.")
+
+        if was_waiting:
+            self.wait_task.cancel()
+            log.debug("Summer AoC stopped during wait task")
+
+        if was_looping:
+            self.loop_task.cancel()  # .cancel() doesn't allow the current iteration to finish
+            log.debug("Summer AoC stopped during loop task")
+
+        self.is_running = False
+        await self.save_event_state()
+        log.info("Summer AoC event stopped")
+        return was_waiting or was_looping
+
+    async def post_puzzle(self) -> None:
+        """Create a thread for the current day's puzzle."""
+        if self.current_day > LAST_DAY:
+            log.error("Attempted to post puzzle after last day, stopping event")
+            await self.stop_event()
+            return
+
+        log.info(f"Posting puzzle for day {self.current_day}")
+        channel: discord.TextChannel = self.bot.get_channel(Channels.summer_aoc_main)
+        thread_starter = await channel.send(
+            f"<@&{Roles.summer_aoc}>",
+            embed=self.get_puzzle_embed(),
+        )
+        await thread_starter.create_thread(name=f"Day {self.current_day} Spoilers")
+
+        self.current_day += 1
+        await self.save_event_state()
+        if self.current_day > LAST_DAY:
+            await self.stop_event()
+
+    def get_info_embed(self) -> discord.Embed:
+        """Generate an embed with info about the event state."""
+        desc = INFO_TEMPLATE.format(
+            is_running=self.is_running,
+            wait_task_active=(self.wait_task is not None) and not self.wait_task.done(),
+            loop_task_active=(self.loop_task is not None) and self.loop_task.is_running(),
```

Comment:
The discord.py tasks module provides a special task for a recurring wait task, which is what we use to wait between posts during the event. To get allow the command to be used before the post-time, a normal async wait is used before the loop starts. So a running event can either be in a state of waiting to make the first post or waiting between posts (wait_task or loop_task).

Ideally the user wouldn't have to know any of that, but I included them in the info embed out of an abundance of caution since it would be necessary information if something went wrong and the info embed should be only staff-facing anyway. If you think they shouldn't be there I don't mind removing them.

Reply:
I've fixed the next puzzle time bug and the bug I mentioned on discord by recording the date of the first post as a reference to calculate the date of the next post.

I also removed the async task info from the embed.